geomstats/geometry/discrete_surfaces.py

```diff
@@ -320,8 +343,567 @@ def surface_metric_matrices(self, point):
+
+    def _inner_product_a0(self, tangent_vec_a, tangent_vec_b, vertex_areas_bp):
+        r"""Compute term of order 0 within the inner-product.
+
+        Denote h and k the tangent vectors a and b respectively.
+        Denote q the base point, i.e. the surface.
+
+        The equation of the inner-product is:
+        :math:`\int_M (G_{a_0} + G_{a_1} + G_{b_1} + G_{c_1} + G_{d_1} + G_{a_2})vol_q`.
+
+        This method computes :math:`G_{a_0} = a_0 <h, k>`,
+        with notations taken from .. [HSKCB2022].
+
+        Parameters
+        ----------
+        tangent_vec_a : array-like, shape=[..., n_vertices, 3]
+            Tangent vector at base point.
+        tangent_vec_b : array-like, shape=[..., n_vertices, 3]
+            Tangent vector at base point.
+        vertex_areas : array-like, shape=[n_vertices, 3]
+            Vertex areas for each vertex of the base_point.
+
+        Returns
+        -------
+        _ : array-like, shape=[...]
+            Term of order 0, and coefficient a0, of the inner-product.
+
+        References
+        ----------
+        .. [HSKCB2022] "Elastic shape analysis of surfaces with second-order
+            Sobolev metrics: a comprehensive numerical framework".
+            arXiv:2204.04238 [cs.CV], 25 Sep 2022.
+        """
+        return self.a0 * gs.sum(
+            vertex_areas_bp
+            * gs.einsum("...bi,...bi->...b", tangent_vec_a, tangent_vec_b),
+            axis=-1,
+        )
+
+    def _inner_product_a1(self, ginvdga, ginvdgb, areas_bp):
+        r"""Compute a1 term of order 1 within the inner-product.
+
+        Denote h and k the tangent vectors a and b respectively.
+        Denote q the base point, i.e. the surface.
+
+        The equation of the inner-product is:
+        :math:`\int_M (G_{a_0} + G_{a_1} + G_{b_1} + G_{c_1} + G_{d_1} + G_{a_2})vol_q`.
+
+        This method computes :math:`G_{a_1} = a_1.g_q^{-1} <dh_m, dk_m>`,
+        with notations taken from .. [HSKCB2022].
+
+        Parameters
+        ----------
+        ginvdga : array-like, shape=[n_faces, 2, 2]
+            Product of the inverse of the surface metric matrices
+            with their differential at a.
+        ginvdgb : array-like, shape=[n_faces, 2, 2]
+            Product of the inverse of the surface metric matrices
+            with their differential at b.
+        areas_bp : array-like, shape=[n_faces,]
+            Areas of the faces of the surface given by the base point.
+
+        Returns
+        -------
+        _ : array-like, shape=[...]
+            Term of order 0, and coefficient a1, of the inner-product.
+
+        References
+        ----------
+        .. [HSKCB2022] "Elastic shape analysis of surfaces with second-order
+            Sobolev metrics: a comprehensive numerical framework".
+            arXiv:2204.04238 [cs.CV], 25 Sep 2022.
+        """
+        return self.a1 * gs.sum(
+            gs.einsum("...bii->...b", gs.matmul(ginvdga, ginvdgb)) * areas_bp,
```

Comment:
use `gs.sum` instead of `einsum`? it seems more clear and probably slightly faster for this case.

Reply:
ah you fully right! I've missed it is only the diagonal.